xfer/model_handler/model_handler.py

```diff
@@ -43,103 +44,485 @@ def __init__(self, module, context_function=mx.context.cpu, num_devices=1, data_
+                for ip in node[consts.INPUTS]:
+                    ip[0] += num_existing_nodes - 1
+
+            # Concatentate nodes list
+            net_symbol_dict[consts.NODES] = net_symbol_dict[consts.NODES] + layer_symbol_dict[consts.NODES][1:]
+
+            # Update attributes of network symbol dictionary
+            net_symbol_dict[consts.HEADS] = self._get_heads(net_symbol_dict[consts.NODES], layer_symbol.name)
+            net_symbol_dict[consts.ARG_NODES] = self._get_arg_nodes(net_symbol_dict[consts.NODES])
+
+        sym = mx.sym.load_json(json.dumps(net_symbol_dict))
         self.update_sym(sym)
         logging.info('Added {} to model top'.format(', '.join(added_layer_names)))
 
-    def add_layer_bottom(self, layer_factory_list):
+    def add_layer_bottom(self, layer_list):
         """
         Add layer to input of model.
         This method requires the entire symbol to be recreated internally.
 
-        :param layer_factory: List of LayerFactory objects to be added to model input.
-        :type layer_factory: list(:class:`LayerFactory`)
+        :param layer_list: List of MxNet symbol layers to be added to model input.
+        :type layer_list: list(:class:`mx.symbol`)
         """
-        # Create data layer
-        symbol_nodes = self._get_symbol_dict()[consts.NODES]
-        data_layer = layer_factory.Data(name=self.data_name)
-        sym = data_layer.create_layer(None)
-        prev_symbols = {self.data_name: sym}
-        id2name = {0: self.data_name}
         added_layer_names = []
-        # Add new layers
-        for id, lf in enumerate(layer_factory_list):
-            self._validate_layer_name(lf.attributes[consts.NAME])
-            if lf.output is True:
-                raise exceptions.ModelError('Cannot add output layer to model bottom ({})'
-                                            .format(lf.attributes[consts.NAME]))
-            sym = lf.create_layer(sym)
-            prev_symbols[lf.attributes[consts.NAME]] = sym
-            # Update id2name with newly added layer
-            id2name[id + 1] = lf.attributes[consts.NAME]
-            added_layer_names.append(lf.attributes[consts.NAME])
-        # Add remaining symbol names to id2name (beginning slice at 1 to skip data node which is already in id2name)
-        for count, node in enumerate(symbol_nodes[1:]):
-            id2name[count + len(layer_factory_list) + 1] = node[consts.LAYER_NAME]
-        # Recreate model
-        sym = self._model_from_nodes(sym, symbol_nodes, elements_offset=len(layer_factory_list),
-                                     prev_symbols=prev_symbols, id2name=id2name, data_name=self.data_name)
+        net_symbol_dict = self._get_symbol_dict(self.symbol)
+
+        for layer_symbol in reversed(layer_list):
+            temp_symbol_dict = copy.deepcopy(net_symbol_dict)
+            added_layer_names.append(layer_symbol.name)
+            layer_symbol_dict = json.loads(layer_symbol.tojson())
```

Comment:
Use the same utility function so that if there is any code change in the future, it can be done in one place.
```suggestion
            layer_symbol_dict = self._get_symbol_dict(layer_symbol)
```